src/types/YXmlFragment.js

```diff
@@ -179,19 +186,60 @@ export class YXmlFragment extends AbstractType {
    *
    * Query support:
    *   - tagname
-   * TODO:
    *   - id
-   *   - attribute
+   *   - attributes
    *
-   * @param {CSS_Selector} query The query on the children.
+   * @param {CSS_Selector|Filters} query The query on the children.
    * @return {YXmlElement|YXmlText|YXmlHook|null} The first element that matches the query or null.
    *
    * @public
    */
   querySelector (query) {
-    query = query.toUpperCase()
-    // @ts-ignore
-    const iterator = new YXmlTreeWalker(this, element => element.nodeName && element.nodeName.toUpperCase() === query)
+    /**
+     * @type {Filters}
+     */
+    let filters = {}
+
+    // Allow passing a string to query the tagname for compatability with DOM api
+    if (typeof query === 'string') {
+      filters.tagname = query
+    } else {
+      filters = query
+    }
+
+    if (filters.tagname) {
+      filters.tagname = filters.tagname.toUpperCase()
+    }
+
+    const iterator = new YXmlTreeWalker(this, element => {
+      // @ts-ignore
+      if (filters.tagname && element.nodeName && element.nodeName.toUpperCase() === filters.tagname) {
+        return true
+      }
+
+      if (filters.id && element._item && element._item.id === filters.id) {
```

Comment:
Could be that there is a more direct way of achieving this that I'm missing?

Reply:
The lamport timestamps should be completely hidden from the user. There are multiple optimizations in Yjs that will lead to issues when using lamport timestamps in this way. 

If you want to retrieve an element by its unique ID (the lamport timestamp), then you should rather call `getItemCleanStart` on the struct store.  This will allow you to retrieve an element in `O(1)` without querying the XML tree.

Reply:
You currently only do an identity check. You probably meant to compare the ids by value (using `compareIDs(id1, id2)`)..

Reply:
Yea, I understand this better now than I did 3 weeks ago. I can remove the `id` querying altogether.